### PR TITLE
feat(client): Add View behavior functions instead of passing around objects.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
This allows for more concise behavior declarations in brokers as well as
provide an abstract type that can be consumed by views in an abstract sense.
A view just calls the behavior with itself and lets the behavior do the rest.
No more adding new code to a bunch of views to add new functionality!

issue #2771

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
Builds on #3109 and should be reviewed afterwards.

**Comment by vladikoff** [comment]
Need rebase

File: app/scripts/views/confirm_reset_password.js [hunk 1]
```diff
@@ -58,8 +58,6 @@ function (Cocktail, ConfirmView, BaseView, Template, p, Session, AuthErrors,
     afterVisible: function () {
       var self = this;
 
-      this.transformLinks();
```
Comment (shane-tomlinson) [inline app/scripts/views/confirm_reset_password.js]: This is taken care of in confirm.js now.

**Comment by shane-tomlinson** [comment]
@vladikoff - rebased!

File: app/scripts/views/base.js [hunk 2]
```diff
@@ -668,6 +668,40 @@ function (Cocktail, _, Backbone, Raven, $, p, AuthErrors,
...
+    /**
+     * Invoke a behavior returned by an auth broker.
+     *
+     * @method invokeBehavior
+     * @param {function} behavior
+     * @return {variant} behavior's return value if behavior is a function,
+     *         otherwise return the behavior.
+     */
+    invokeBehavior: function (behavior) {
+      if (typeof behavior === 'function') {
```
Comment (vladikoff) [inline app/scripts/views/base.js]: nit: could we use http://underscorejs.org/#isFunction ? _ is already available in the file

File: app/scripts/views/confirm_reset_password.js [hunk 3]
```diff
@@ -181,6 +176,7 @@ function (Cocktail, ConfirmView, BaseView, Template, p, Session, AuthErrors,
 
     _finishPasswordResetDifferentBrowser: function () {
       var self = this;
+
```
Comment (vladikoff) [inline app/scripts/views/confirm_reset_password.js]: nit: no need for line break

File: app/scripts/views/permissions.js [hunk 4]
```diff
@@ -88,13 +86,9 @@ function (Cocktail, FormView, Template, p, BackMixin, ServiceMixin) {
...
-      return self.broker.afterSignIn(account)
-        .then(function (result) {
-          if (! (result && result.halt)) {
-            self.navigate('settings');
-          }
-
-          return result;
+      return self.invokeBrokerMethod('afterSignIn', account)
+        .then(function () {
+          self.navigate('settings');
```
Comment (vladikoff) [inline app/scripts/views/permissions.js]: :+1:

**Comment by vladikoff** [comment]
r+ Looks good, just a few nits, feel free to merge after (if you decide to update)

File: app/scripts/views/base.js (hunk 2 above)
Comment (shane-tomlinson) [inline app/scripts/views/base.js]: Yes, yes I should use that. :+1:

**Pull request merged**